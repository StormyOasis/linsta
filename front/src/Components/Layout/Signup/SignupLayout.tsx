import React from "react";
import { styled } from "styled-components";

import * as styles from '../../../Components/Common/CombinedStyling';

import Theme from "../../../Components/Themes/Theme";
import SignupMainForm from "../../../Components/Layout/Signup/Flow/SignupMainForm";
import BirthdayForm from "../../../Components/Layout/Signup/Flow/BirthdayForm";
import ConfirmationCodeForm from "../../../Components/Layout/Signup/Flow/ConfirmationCodeForm";
import StyledLink from "../../../Components/Common/StyledLink";
import { Div } from "../../../Components/Common/CombinedStyling";

type SignupLayoutProps = {
    page?: number | null
};

type SignupState = {
    userName: string;
    fullName: string;
    emailOrPhone: string;
    password: string;
    confirmationCode: string;

    month: number;
    day: number;
    year: number;

    userName_valid: boolean;
    fullName_valid: boolean;
    emailOrPhone_valid: boolean;
    password_valid: boolean;
    year_valid: boolean;
    confirmationCode_valid: boolean;

    signupPage: number;
    showBirthdayModal: boolean;
};

const SignupLayoutWrapper = styled.main<any>`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
`;

export default class SignupLayout extends React.Component<
    SignupLayoutProps,
    SignupState
> {
    constructor(props: SignupLayoutProps) {
        super(props);

        this.state = {
            userName: "",
            fullName: "",
            emailOrPhone: "",
            password: "",
            confirmationCode: "",

            month: 1,
            day: 1,
            year: new Date().getFullYear(),

            userName_valid: false,
            fullName_valid: false,
            emailOrPhone_valid: false,
            password_valid: false,
            confirmationCode_valid: false,
            year_valid: false,
            signupPage: props.page ? props.page : 0,
            showBirthdayModal: false
        };
    }

    changePage = (changeValue: number) => {
        let newPage = this.state.signupPage + changeValue;
        if (newPage < 0) {
            newPage = 0;
        } else if (newPage > 2) {
            newPage = 2;
        }

        this.setState({ signupPage: newPage }, () => { this.forceUpdate() });
    }

    changeState = (key: string, value: any) => {
        this.setState({
            ...this.state,
            [key]: value
        });
    }

    handleFormChange = (key: string, value: string, valid: boolean) => {
        this.setState({
            ...this.state,
            [key]: value,
            [key + "_valid"]: valid,
        });        
    }

    renderPage = () => {
        if (this.state.signupPage == 0) {
            return <SignupMainForm 
                emailOrPhone={this.state.emailOrPhone} 
                fullName={this.state.fullName} 
                userName={this.state.userName} 
                password={this.state.password}
                emailOrPhone_valid={this.state.emailOrPhone_valid}
                fullName_valid={this.state.fullName_valid}
                userName_valid={this.state.userName_valid}
                password_valid={this.state.password_valid}
                changePage={this.changePage}
                handleFormChange={this.handleFormChange}
                />
        } else if (this.state.signupPage == 1) {
            return <BirthdayForm 
                changePage={this.changePage} 
                changeState={this.changeState}                 
                handleFormChange={this.handleFormChange} 
                showBirthdayModal={this.state.showBirthdayModal} 
                date_valid={this.state.year_valid} 
                day={this.state.day} 
                month={this.state.month} 
                year={this.state.year} />;
        } else if (this.state.signupPage == 2) {
            return <ConfirmationCodeForm 
                confirmationCode={this.state.confirmationCode} 
                confirmationCode_valid={this.state.confirmationCode_valid} 
                emailOrPhone={this.state.emailOrPhone} 
                fullName={this.state.fullName} 
                userName={this.state.userName} 
                password={this.state.password}
                day={this.state.day} 
                month={this.state.month} 
                year={this.state.year}                 
                handleFormChange={this.handleFormChange} 
                changePage={this.changePage}/>
        }
        return <></>;
    }

    override render() {
        return (
            <Theme>
                <SignupLayoutWrapper role="main">
                    <div className={styles.default.innerDiv1}>
                        <div className={styles.default.innerDiv2}>
                            <div className={styles.default.signupBox}>
                                {this.renderPage()}
                            </div>
                            <div
                                className={styles.default.signupBox}
                                style={{ padding: "20px 0" }}>
                                    Have an account? <StyledLink to="/login">Log In</StyledLink>
                            </div>
                        </div>
                    </div>
                    <Div $margin="auto">
                        <StyledLink to="/about" styleOverride={{fontSize: ".925em", fontWeight: 600}}>                            
                            About
                        </StyledLink>                     
                    </Div>                    
                </SignupLayoutWrapper>
            </Theme>
        );
    }
}
