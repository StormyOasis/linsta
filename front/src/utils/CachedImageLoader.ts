import { clear, del, get, set } from 'idb-keyval';
import { blobToBase64 } from "./utils";

export const loadImageCached = async (id: string, url: string): Promise<string> => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // Image not in the cache so have to fetch it
        try {
            const result = await fetch(url);
            if(result.status === 200) {
                const blob = await result.blob();
                const b64:string = await blobToBase64(blob) as string;

                // add the base64 data into the cache and return it
                await set(id, b64);

                return b64;
            }
        } catch(err) {
            console.log(err);
            return url; //default to what was passed in
        }

    } else {
        // Image is in the cache already as a base64 string, use that as is
        return value;
    }

    return url; //default to what was passed in
}

export const setImage = async (id: string, data: string) => {     
    await set(id, data);
}

export const deleteFromCache = async (id: string) => {
    // First try to get the image from the cache
    const value = await get(id);

    if(value === undefined) {
        // key not present in cache so do nothing
        return;
    }

    // delete from the cache
    await del(id);
}

export const clearCache = async () => {
    await clear();
}