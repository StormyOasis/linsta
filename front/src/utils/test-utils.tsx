import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { RootState } from "../Components/Redux/redux";
import { authInitialState } from '../Components/Redux/slices/auth.slice';
import { modalInitialState } from '../Components/Redux/slices/modals.slice';
import { profileInitialState } from '../Components/Redux/slices/profile.slice';
import { miscInitialState } from '../Components/Redux/slices/misc.slice';
import { configureStore } from '@reduxjs/toolkit';

import Theme from '../Components/Themes/Theme';

export const mockInitialState = (overrides?: Partial<RootState>): RootState => ({
    auth: { ...authInitialState, ...(overrides?.auth ?? {}) },
    modal: { ...modalInitialState, ...(overrides?.modal ?? {}) },
    profile: { ...profileInitialState, ...(overrides?.profile ?? {}) },
    misc: { ...miscInitialState, ...(overrides?.misc ?? {}) },
});

type ExtendedRenderOptions = {
    store?: ReturnType<typeof configureStore>;
    preloadedState?: any;
    route?: string; // ‚Üê allow overriding the initial route
} & Omit<RenderOptions, "queries">;

const dummyAuthReducer = (state = { user: null, isLoading: false, error: null }, _action: any) => state;
const dummyModalReducer = (state = { isOpen: false }, _action: any) => state;
const dummyProfileReducer = (state = { name: '', email: '' }, _action: any) => state;
const dummyMiscReducer = (state = { someData: null }, _action: any) => state;

const renderWithStore = (
    ui: ReactElement,
    {
        preloadedState,
        store = configureStore({
            reducer: {
                auth: dummyAuthReducer,
                modal: dummyModalReducer,
                profile: dummyProfileReducer,
                misc: dummyMiscReducer,
            },
            preloadedState,
        }),
        route = "/",
        ...renderOptions
    }: ExtendedRenderOptions = {}
) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
        return (
            <Provider store={store}>
                <MemoryRouter initialEntries={[route]}>
                    <Theme>{children}</Theme>
                </MemoryRouter>
            </Provider>
        );
    };

    return {
        ...render(ui, { wrapper: Wrapper, ...renderOptions }),
        store,
    };
};

export * from "@testing-library/react";
export { renderWithStore };
export type { ExtendedRenderOptions };