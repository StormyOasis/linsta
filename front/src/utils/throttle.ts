import { useRef } from 'react';

const useThrottle = (callback: Function, delay: number) => {
    const lastCall = useRef(0);

    return (...args: any[]) => {
        const now = new Date().getTime();
        if (now - lastCall.current >= delay) {
            lastCall.current = now;
            callback(...args);
        }
    };
};

export default useThrottle;