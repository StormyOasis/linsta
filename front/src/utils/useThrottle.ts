import { useRef, useCallback } from 'react';

/**
 * React hook for throttling a callback function.
 * The callback will only be invoked at most once every `delay` milliseconds.
 *
 * @param callback - The function to throttle.
 * @param delay - The minimum delay (ms) between calls.
 * @returns A throttled version of the callback.
 */
const useThrottle = <T extends (...args: any[]) => void>(callback: T, delay: number) => {
    const lastCall = useRef<number>(0);

    return useCallback(
        (...args: Parameters<T>) => {
            const now = performance.now();
            if (now - lastCall.current >= delay) {
                lastCall.current = now;
                callback(...args);
            }
        },
        [callback, delay]
    );
};

export default useThrottle;