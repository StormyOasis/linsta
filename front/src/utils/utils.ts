import sanitizeHtml from 'sanitize-html';

import { HistoryType, Post, Profile } from "../api/types";
import { postSetFollowStatus, postUpdatePost } from '../api/ServiceController';
import { DEFAULT_PFP } from '../api/config';

/**
 * Utility object for managing navigation and history state.
 */
export const historyUtils: HistoryType = {
    navigate: null,
    location: null,
    isServer: true
}

// Regular expressions for validation
const validatePasswordRegex: RegExp = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@.#$!%*?&^_])[A-Za-z\d@.#$!%*?&_]{8,32}$/;
const phoneRegex: RegExp = /(?:([+]\d{1,4})[-.\s]?)?(?:[(](\d{1,3})[)][-.\s]?)?(\d{1,4})[-.\s]?(\d{1,4})[-.\s]?(\d{1,9})/g;
const emailRegex: RegExp = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/;

/**
 * Checks if a string is a hashtag (starts with '#').
 */
export const isHashtag = (text: string) => text.startsWith('#');

/**
 * Checks if a string is a mention (starts with '@').
 */
export const isMention = (text: string) => text.startsWith('@');

export const validatePassword = (value: string): boolean => {
    return validatePasswordRegex.test(value);
};

export const validateEmailPhone = (value: string): boolean => {
    if (value == null) {
        return false;
    }

    return emailRegex.test(value) || phoneRegex.test(value);
};

/**
 * Checks if a full name contains at least two words.
 */
export const validateFullName = (value: string): boolean => {
    return value?.trim().split(" ").length > 1;
};

/**
 * Validates a URL string. Empty or undefined is considered valid.
 */
export const validateUrl = (text: string | undefined): boolean => {
    if (text == null || text.length === 0) {
        return true; //No url entered should be considered valid
    }

    try {
        new URL(text);
        return true;
    } catch (e) {
        return false;
    }
}

// Supported video file extensions
const validVideoExtensions: string[] = ["avif", "ogm", "wmv", "mpg", "webm", "ogv", "mov", "asx", "mpeg", "mp4", "m4v", "avi"];
/**
 * Checks if a file path has a valid video extension.
 * @throws Error if the path is invalid.
 */
export const isVideoFileFromPath = (path: string): boolean => {
    if (!path || path.trim().length < 4) {
        throw new Error("Invalid path");
    }

    const ext = path.toLowerCase().substring(path.lastIndexOf(".") + 1);

    return validVideoExtensions.includes(ext);
}

/**
 * Checks if a MIME type string indicates a video.
 * @throws Error if the type is invalid.
 */
export const isVideoFileFromType = (type: string): boolean => {
    if (type == null || type.trim().length < 4) {
        throw new Error("Invalid type");
    }

    return type.toLowerCase().includes("video");
}

/**
 * Converts a base64 string to a File blob object.
 */
export const base64ToBlob = (base64String: string, outFileName: string): File | null => {
    if (base64String === null || base64String.length === 0) {
        return null;
    }

    const contentType = base64String.substring(5, base64String.indexOf(';'));
    const data = base64String.substring(base64String.indexOf(",") + 1);

    return base64ToBlobEx(data, contentType, outFileName);
}

/**
 * Converts a base64 string (without prefix) to a File object.
 */
export const base64ToBlobEx = (base64String: string, contentType: string, outFileName: string): File => {
    const sliceSize = 512;
    const byteCharacters = atob(base64String);
    const byteArrays = [];

    for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
        const slice = byteCharacters.slice(offset, offset + sliceSize);
        const byteNumbers = new Array(slice.length);

        for (let i = 0; i < slice.length; i++) {
            byteNumbers[i] = slice.charCodeAt(i);
        }

        byteArrays.push(new Uint8Array(byteNumbers));
    }

    return new File(byteArrays, outFileName, { type: contentType });
}

/**
 * Converts a Blob to a base64 string.
 */
export const blobToBase64 = async (blob: any): Promise<string | ArrayBuffer | null> => {
    blob = await createBlob(blob);

    return new Promise((resolve, _reject) => {
        const reader = new FileReader();
        reader.onloadend = () => resolve(reader.result);
        reader.readAsDataURL(blob);
    });
}

/**
 * Fetches a resource as a Blob.
 */
export const createBlob = async (url: string): Promise<Blob | null> => {
    try {
        const response = await fetch(url);
        const data = await response.blob();
        return data;
    } catch (err) {
        console.log(err);
    }
    return null;
}

/**
 * Extracts the MIME type from a base64 data URL.
 */
export const extractMimeTypeFromBase64 = (data: string): string | null => {
    if (data == null || data.length === 0) {
        return null;
    }

    return data.substring(5, data.indexOf(';') + 1);
}

/**
 * Extracts a frame from a video element as a JPEG blob URL.
 */
export const extractFrameFromVideo = async (video: HTMLVideoElement): Promise<string | null> => {
    if (video == null) {
        return null;
    }

    const canvas = document.createElement("canvas");
    const ctx = canvas.getContext('2d');

    if (ctx == null) {
        return null;
    }

    canvas.height = video.videoHeight || parseInt(video.style.height);
    canvas.width = video.videoWidth || parseInt(video.style.width);

    ctx.drawImage(video, 0, 0);

    return new Promise((resolve, _reject) => {
        canvas.toBlob((file: any) =>
            resolve(URL.createObjectURL(file))
            , 'image/jpeg')
    })
}

/**
 * Returns a human-readable difference between now and the given date.
 */
export const dateDiff = (dateTime: Date): string => {
    if (dateTime == null) {
        return "now";
    }

    const now = new Date().getTime();
    const dateTimeMilli = new Date(dateTime).getTime();
    const diff = new Date(now - dateTimeMilli);
    let diffMilli = diff.getTime();

    const days = Math.floor(diffMilli / 1000 / 60 / 60 / 24);
    diffMilli = diffMilli - (days * 1000 * 60 * 60 * 24);

    const weeks = Math.floor(days / 7.0);
    const hours = Math.floor(diffMilli / 1000 / 60 / 60);
    diffMilli = diffMilli - (hours * 1000 * 60 * 60);
    const mins = Math.floor(diffMilli / 1000 / 60);

    if (weeks >= 1) {
        return weeks + "w";
    }

    if (days >= 1) {
        return days + "d";
    }

    if (hours >= 1) {
        return hours + "h";
    }

    if (mins >= 1) {
        return mins + "m";
    }

    return "now";
}

/**
 * Formats a Date object as a readable string.
 */
export const getDateAsText = (date: Date): string => {
    return new Date(date).toLocaleDateString('en-us', { year: "numeric", month: "long", day: "numeric" });
}

/**
 * Finds a post by ID from a list of posts.
 */
export const getPostFromListById = (postId: string, posts: Post[]): Post => {
    return posts.find(post => post.postId === postId) ?? ({} as any);
}

/**
 * Checks if a post is liked by a given user.
 */
export const isPostLiked = (userName: string, post: Post): boolean => {
    return !!(post?.global?.likes?.some((like: any) => like.userName === userName));
}

/**
 * Toggles the liked state of a post for a user (Only on the front end, no backend call here).
 */
export const togglePostLikedState = (userName: string, userId: string, post: Post): Post => {
    if (post == null || userName == null || userId == null) {
        return post;
    }

    if (post.global.likes == null) {
        post.global.likes = [];
    }

    const index = post.global.likes.findIndex((value: any) => value.userName === userName);
    if (index === -1) {
        // Username is not in the post's like list, so add it
        post.global.likes.push({ userName, userId });
    } else {
        // Remove the username from the post's like list
        post.global.likes.splice(index, 1);
    }

    return post;
}

/**
 * Sanitizes text for safe HTML display and also returns a plain text version.
 */
export const getSanitizedText = (text: string): [string, string] => {
    if (text == null || text.length === 0)
        return [text, text];

    const textHtml = sanitizeHtml(text, {
        allowedTags: ['b', 'i', 'em', 'strong', 'a', 'br', 'sub', 'sup'],
        allowedAttributes: {
            'a': ['href']
        },
        allowedIframeHostnames: []
    }).trim();

    const textStripped = sanitizeHtml(text, {
        allowedTags: [],
        allowedAttributes: {},
        allowedIframeHostnames: []
    }).trim();

    return [textHtml, textStripped];
}

/**
 * Checks if an element with the given ID is vertically overflowed.
 */
export const isOverflowed = (id: string): boolean => {
    const element = document.getElementById(id);

    if (element == null) {
        return false;
    }

    return element.offsetHeight < element.scrollHeight;
}

/**
 * Follows or unfollows a user by calling the backend service.
 */
export const followUser = async (userId: string, followUserId: string, shouldFollow: boolean): Promise<boolean> => {
    if (userId === followUserId) {
        return true;
    }

    const data = {
        follow: shouldFollow,
        userId: userId,
        followId: followUserId
    };

    const result = await postSetFollowStatus(data);

    return result.status === 200;
}

/**
 * Returns the profile picture URL from a Profile object, or a default if missing.
 */
export const getPfpFromProfile = (profile: Profile | null): string => {
    return profile?.pfp?.length ? profile.pfp : DEFAULT_PFP;
}

/**
 * Returns the profile picture URL from a Post object, or a default if missing.
 */
export const getPfpFromPost = (post: Post): string => {
    if (post.user.pfp == null || post.user.pfp.length === 0) {
        return DEFAULT_PFP;
    }

    return post.user.pfp;
}

/**
 * Splits a full name into first, middle, and last names.
 */
export const splitFullName = (fullName: string): { firstName: string, middleNames: string, lastName: string } => {
    const names = fullName.trim().split(' ');
    const firstName = names[0];
    const lastName = names[names.length - 1];
    const middleNames = names.slice(1, names.length - 1).join(' ');

    return { firstName, middleNames, lastName };
}

/**
 * Updates fields on a post and calls onClose with the result.
 */
export const updatePostFields = async (post: Post, fieldsToUpdate: { key: string, value: any }[], onClose: (data: any) => void) => {
    if (fieldsToUpdate == null || fieldsToUpdate.length === 0) {
        return;
    }

    try {
        const newPost: Post = await updatePost(post, fieldsToUpdate);
        // Close the modal
        onClose({ isCommited: true, post: newPost, isDeleted: false });
    } catch (err) {
        console.error(err)
    }
}


/**
 * Updates a post with the given fields and returns the updated post.
 */
export const updatePost = async (post: Post, fieldsToUpdate: { key: string, value: any }[]): Promise<Post> => {
    if (fieldsToUpdate == null || fieldsToUpdate.length === 0) {
        return post;
    }

    try {
        // build the query from the given fieldsToUpdate array
        let fields: any = null;
        if (fieldsToUpdate.length > 0) {
            fields = {};
            fieldsToUpdate.forEach((field) => {
                fields[field.key] = field.value;
            });
        }

        // Call the update post service
        const results = await postUpdatePost({ postId: post.postId, fields });
        if (results == null || results.status !== 200) {
            throw new Error("Error updating post");
        }

        // update the local returned post adding the fields not returned by ES
        results.data.postId = post.postId
        results.data.user.pfp = post.user.pfp;
        results.data.commentCount = post.global.commentCount;
        if (post?.global?.likes != null) {
            results.data.global.likes = [...post?.global?.likes];
        }

        return results.data as Post;
    } catch (err) {
        console.error(err)
    }
    return post;
}

/**
 * Stores a search query (string or Profile) in localStorage under "recentSearches".
 * Prevents duplicates based on string value or Profile.profileId.
 * @param value - The search query to store.
 * @returns The updated array of stored queries.
 */
export const storeSearchQueries = (value: string | Profile): (string | Profile)[] => {
    const key: string = "recentSearches";
    try {
        const storedData = localStorage.getItem(key);
        const parsedData = storedData ? JSON.parse(storedData) : [];

        // Check if value is a duplicate
        const isDuplicate: boolean = parsedData.some((item: string | Profile) => {
            if (typeof item === 'string') {
                return item === value;
            } else {
                return item.profileId === (value as Profile).profileId;
            }
        })

        if (isDuplicate) {
            return parsedData;
        }

        parsedData.push(value);

        localStorage.setItem(key, JSON.stringify(parsedData));
        return parsedData;
    } catch (err) {
        console.error(err);
    }
    return [];
}

/**
 * Removes a search query (string or Profile) from localStorage "recentSearches".
 * @param value - The search query to remove.
 * @returns The updated array of stored queries.
 */
export const removeStoredSearchQuery = (value: string | Profile): (string | Profile)[] => {
    const key: string = "recentSearches";
    let queries: (string | Profile)[] = [];
    try {
        queries = getStoredSearchQueries();
        const filteredQueries = queries.filter((item: string | Profile) => {
            if (typeof item === 'string') {
                return item !== value;
            } else {
                return item.profileId !== (value as Profile).profileId;
            }
        });
        localStorage.setItem(key, JSON.stringify(filteredQueries));
        return filteredQueries;
    } catch (err) {
        return queries;
    }
}

/**
 * Retrieves all stored search queries (strings or Profiles) from localStorage.
 * @returns An array of stored queries.
 */
export const getStoredSearchQueries = (): (string | Profile)[] => {
    const key: string = "recentSearches";
    const item = localStorage.getItem(key);

    if (!item) {
        return [];
    }

    try {
        return JSON.parse(item);
    } catch (err) {
        return [];
    }
}

/**
 * Builds a space-separated string of user names from selected Profile objects.
 * @param selectedProfiles - A record of Profile objects keyed by string.
 * @returns A string of user names separated by spaces.
 */
export const buildCollabSearchText = (selectedProfiles: Record<string, Profile>): string => {
    if (!selectedProfiles) {
        return "";
    }

    // Clear out the text input if there's no results
    let userNames: string = "";

    Object.values(selectedProfiles).forEach((profile: Profile) => {
        userNames = `${userNames}${profile.userName} `;
    });

    return userNames.trim();
}