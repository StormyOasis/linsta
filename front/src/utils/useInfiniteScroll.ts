import React, { useEffect, useRef } from 'react';
import useThrottle from './useThrottle';

/**
 * Type for the async load more function.
 */
type AsyncLoadMoreFunction = () => Promise<void>;

/**
 * Custom hook for infinite scrolling.
 * Attaches a throttled scroll event listener to a container and triggers `loadMore`
 * when the user scrolls near the bottom.
 *
 * @param loadMore - Async function to load more items.
 * @param isLoading - Whether a load is currently in progress.
 * @param containerRef - Ref to the scrollable container element.
 */
const useInfiniteScroll = (loadMore: AsyncLoadMoreFunction, 
    isLoading: boolean, 
    childRef: React.MutableRefObject<HTMLDivElement | null>) => {
        
    const hasScrolled = useRef<boolean>(false);  // Track if the user has scrolled

    // Throttled scroll handler
    const throttledHandleScroll = useThrottle(async () => {
        if (typeof window !== 'undefined' && childRef.current) {
            const element = childRef.current as HTMLElement;
            const currentScroll = window.innerHeight + element.scrollTop;

            if (currentScroll + 256 >= element.scrollHeight && !isLoading && !hasScrolled.current) {
                hasScrolled.current = true;  // Mark as scrolled at least once
                await loadMore();
            }
        }
    }, 200);

    useEffect(() => {
        const childElement = childRef.current;
        if (childElement) {
            childElement.addEventListener('scroll', throttledHandleScroll);
        }

        // Clean up event listener on unmount
        return () => {
            if (childElement) {
                childElement.removeEventListener('scroll', throttledHandleScroll);
            }
        };
    }, [throttledHandleScroll, childRef]);

    // Reset trigger flag when loading finishes
    useEffect(() => {
        if (!isLoading) {
            hasScrolled.current = false;
        }
    }, [isLoading]);
}

export default useInfiniteScroll;