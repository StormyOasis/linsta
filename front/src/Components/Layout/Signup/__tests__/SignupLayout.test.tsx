import React from "react";
import { screen } from "@testing-library/react";
import SignupLayout from "../SignupLayout";
import { renderWithStore } from "../../../../utils/test-utils";
import '@testing-library/jest-dom';

// Mocks
jest.mock("../../../Common/CombinedStyling", () => ({
    __esModule: true,
    ...jest.requireActual("../../../Common/CombinedStyling"),
    default: {
        innerDiv1: "innerDiv1",
        innerDiv2: "innerDiv2",
        signupBox: "signupBox"
    }
}));
jest.mock("../../../Themes/Theme", () => ({
    __esModule: true,
    default: ({ children }: any) => <div data-testid="mock-theme">{children}</div>
}));
jest.mock("../Flow/SignupMainForm", () => (props: any) => <div data-testid="signup-main-form">{JSON.stringify(props)}</div>);
jest.mock("../Flow/BirthdayForm", () => (props: any) => <div data-testid="birthday-form">{JSON.stringify(props)}</div>);
jest.mock("../Flow/ConfirmationCodeForm", () => (props: any) => <div data-testid="confirmation-code-form">{JSON.stringify(props)}</div>);
jest.mock("../../../Common/StyledLink", () => (props: any) => (
    <a data-testid="styled-link" href={props.to}>{props.children}</a>
));

describe("SignupLayout", () => {
    it("renders SignupMainForm on page 0", () => {
        const r = renderWithStore(<SignupLayout page={0} />);
        expect(screen.getByTestId("signup-main-form")).toBeInTheDocument();
        expect(screen.getByTestId("styled-link")).toHaveAttribute("href", "/login");
        expect(r.container).toMatchSnapshot();
    });

    it("renders BirthdayForm on page 1", () => {
        const r = renderWithStore(<SignupLayout page={1} />);
        expect(screen.getByTestId("birthday-form")).toBeInTheDocument();
        expect(screen.getByTestId("styled-link")).toHaveAttribute("href", "/login");
        expect(r.container).toMatchSnapshot();
    });

    it("renders ConfirmationCodeForm on page 2", () => {
        const r = renderWithStore(<SignupLayout page={2} />);
        expect(screen.getByTestId("confirmation-code-form")).toBeInTheDocument();
        expect(screen.getByTestId("styled-link")).toHaveAttribute("href", "/login");
        expect(r.container).toMatchSnapshot();
    });

    it("defaults to SignupMainForm if no page prop is given", () => {
        const r = renderWithStore(<SignupLayout />);
        expect(screen.getByTestId("signup-main-form")).toBeInTheDocument();
        expect(r.container).toMatchSnapshot();
    });
});