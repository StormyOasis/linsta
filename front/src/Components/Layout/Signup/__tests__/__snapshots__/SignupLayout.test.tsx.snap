// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupLayout defaults to SignupMainForm if no page prop is given 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-theme"
    >
      <main
        class="SignupLayout__SignupLayoutWrapper-sc-17nyy42-0 exFDKZ"
        role="main"
      >
        <div
          class="innerDiv1"
        >
          <div
            class="innerDiv2"
          >
            <div
              class="signupBox"
            >
              <div
                data-testid="signup-main-form"
              >
                {"emailOrPhone":"","fullName":"","userName":"","password":"","emailOrPhone_valid":false,"fullName_valid":false,"userName_valid":false,"password_valid":false}
              </div>
            </div>
            <div
              class="signupBox"
              style="padding: 20px 0px;"
            >
              Have an account? 
              <a
                data-testid="styled-link"
                href="/login"
              >
                Log In
              </a>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;

exports[`SignupLayout renders BirthdayForm on page 1 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-theme"
    >
      <main
        class="SignupLayout__SignupLayoutWrapper-sc-17nyy42-0 exFDKZ"
        role="main"
      >
        <div
          class="innerDiv1"
        >
          <div
            class="innerDiv2"
          >
            <div
              class="signupBox"
            >
              <div
                data-testid="birthday-form"
              >
                {"showBirthdayModal":false,"date_valid":false,"day":1,"month":1,"year":2025}
              </div>
            </div>
            <div
              class="signupBox"
              style="padding: 20px 0px;"
            >
              Have an account? 
              <a
                data-testid="styled-link"
                href="/login"
              >
                Log In
              </a>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;

exports[`SignupLayout renders ConfirmationCodeForm on page 2 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-theme"
    >
      <main
        class="SignupLayout__SignupLayoutWrapper-sc-17nyy42-0 exFDKZ"
        role="main"
      >
        <div
          class="innerDiv1"
        >
          <div
            class="innerDiv2"
          >
            <div
              class="signupBox"
            >
              <div
                data-testid="confirmation-code-form"
              >
                {"confirmationCode":"","confirmationCode_valid":false,"emailOrPhone":"","fullName":"","userName":"","password":"","day":1,"month":1,"year":2025}
              </div>
            </div>
            <div
              class="signupBox"
              style="padding: 20px 0px;"
            >
              Have an account? 
              <a
                data-testid="styled-link"
                href="/login"
              >
                Log In
              </a>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;

exports[`SignupLayout renders SignupMainForm on page 0 1`] = `
<div>
  <div
    data-testid="mock-theme"
  >
    <div
      data-testid="mock-theme"
    >
      <main
        class="SignupLayout__SignupLayoutWrapper-sc-17nyy42-0 exFDKZ"
        role="main"
      >
        <div
          class="innerDiv1"
        >
          <div
            class="innerDiv2"
          >
            <div
              class="signupBox"
            >
              <div
                data-testid="signup-main-form"
              >
                {"emailOrPhone":"","fullName":"","userName":"","password":"","emailOrPhone_valid":false,"fullName_valid":false,"userName_valid":false,"password_valid":false}
              </div>
            </div>
            <div
              class="signupBox"
              style="padding: 20px 0px;"
            >
              Have an account? 
              <a
                data-testid="styled-link"
                href="/login"
              >
                Log In
              </a>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
